File: Difficulty: Medium/Count Number of Nodes in a Binary Tree/count-number-of-nodes-in-a-binary-tree.cpp
// User function Template for C++

class Solution {
  public:
    int countNodes(Node* root) {
        // Write your code here
        if(root==NULL) return 0;
        
        int lh = leftheight(root);
        int rh = rightheight(root);
        
        if(lh==rh) return (1<<lh)-1;
        return 1+countNodes(root->left)+countNodes(root->right);
    }
    int leftheight(Node* root){
        int h=0;
        while(root){
            h++;
            root = root->left;
        }
        return h;
    }
    int rightheight(Node* root){
        int h=0;
        while(root){
            h++;
            root = root->right;
        }
        return h;
    }
    
    
};

File: Difficulty: Medium/Count Number of Nodes in a Binary Tree/count-number-of-nodes-in-a-binary-tree.py
#User function Template for python3

'''
# Node Class:
class Node:
    def init(self, val):
        self.data = val
        self.left = None
        self.right = None
'''
class Solution:
    def leftheight(self,root):
        h=0
        while root:
            h+=1
            root=root.left
        return h
    
    def rightheight(self,root):
        h=0
        while root:
            h+=1
            root=root.right
        return h
        
    def countNodes(self, root):
        #code here
        if root is None:
            return 0
        
        lh = self.leftheight(root)
        rh = self.rightheight(root)
        
        if lh==rh:
            return (1<<lh)-1
        return 1+self.countNodes(root.left) + self.countNodes(root.right)
            